File: Sorting/sorting_algorithms.c
#include <stdio.h>
#include <stdlib.h>

void swap(int *arr, int i, int j){
	int tmp;

	tmp = arr[i];
	arr[i] = arr[j];
	arr[j] = tmp;
}

void selection_sort(int *num_arr, int arr_len){
	int min_idx;

	for (int idx = 0; idx < arr_len; idx++)
	{
		min_idx = idx;
		for (int i = idx; i < arr_len; i++)
		{
			if (num_arr[i] <= num_arr[min_idx])
				min_idx = i;
		}
		swap(num_arr, idx, min_idx);
	}
}

void bubble_sort(int *num_arr, int arr_len){
	int j;
	while (arr_len > 0)
	{
		j = 0;
		while (j < arr_len - 1)
		{
			if (num_arr[j] > num_arr[j + 1])
				swap(num_arr, j, j+1);
			j++;
		}
		arr_len--;
	}
}

void insertion_sort(int *num_arr, int arr_len){

	int i, j, key;

	if (arr_len == 1)
		return;

	for (i = 1; i <= arr_len; i++){
		key = num_arr[i];
		for (j = i-1; j >= 0 && num_arr[j]>key; j--){
			num_arr[j + 1] = num_arr[j];
		}
		num_arr[j + 1] = key;
	}
}

int sorted[10];
void merge(int *list, int left, int mid, int right){
  int i, j, k, l;
  i = left;
  j = mid+1;
  k = left;

  while(i<=mid && j<=right){
    if(list[i]<=list[j])
      sorted[k++] = list[i++];
    else
      sorted[k++] = list[j++];
  }

  while (i <= mid)
	  sorted[k++] = list[i++];

  while (j <= right)
	  sorted[k++] = list[j++];

  for(l=left; l<=right; l++)
    list[l] = sorted[l];
}

void merge_sort(int *num_arr, int left, int right){
	int mid;
	if (left < right) {
		mid = (left + right) / 2; 
		merge_sort(num_arr, left, mid);
		merge_sort(num_arr, mid+1, right);
		merge(num_arr, left, mid, right);
	}
}


void quick_sort(int *num_arr, int L, int R){
	int left = L, right = R;
	int pivot = num_arr[(L + R) / 2];

	while (left <= right)
	{
		while (num_arr[left] < pivot)
			left++;
		while (num_arr[right] > pivot)
			right--;
		if (left <= right)
		{
			swap(num_arr, left, right);
			left++;
			right--;
		}
		if (L < right) //아직 끝까지 정렬하지 못함
			quick_sort(num_arr, L, right);
		if (left < R)
			quick_sort(num_arr, left, R);
	}
}


int main() {
	int n = 10;
	int *num_arr = (int *)malloc(sizeof(int) * 10);
	for (int i = 0; i < 10; i++){
		num_arr[i] = rand() % 100;
	}

	//selection_sort(num_arr, n);
	//insertion_sort(num_arr, n);
	//merge_sort(num_arr, 0, n-1);
	//bubble_sort(num_arr, n);
	quick_sort(num_arr, 0, n-1);

	for (int i = 0; i < n; i++)
		printf("%d ", num_arr[i]);
}

File: Sorting/sorting_algorithms.py
import random
from turtle import left

def bubble_sort(num_list):
    for i in range(len(num_list)-1):
        swap = False
        for j in range(len(num_list)-i-1):
            if num_list[j] > num_list[j+1]:
                num_list[j], num_list[j+1] = num_list[j+1], num_list[j]
                swap = True

        if swap == False:
            break

    return num_list

def insertion_sort(num_list):
    for i in range(1, len(num_list)):
        for j in range(i):
            if num_list[j] > num_list[i]:
                num_list[i], num_list[j] = num_list[j], num_list[i]
            else:
                break
    return num_list


def selection_sort(num_list):
    for i in range(len(num_list) - 1):
        min_idx = i
        for j in range(i, len(num_list)):
            if num_list[j] < num_list[min_idx]:
                min_idx = j
        num_list[i], num_list[min_idx] = num_list[min_idx], num_list[i]
    return num_list


def merge_sort(num_list):
    i, j, k = 0, 0, 0
    if len(num_list) == 1:
        return num_list


    mid = len(num_list) // 2
    left_list = num_list[:mid]
    right_list = num_list[mid:]
    merge_sort(left_list)
    merge_sort(right_list)

    while (i < len(left_list) and j < len(right_list)):
        if (left_list[i] <= right_list[j]):
            num_list[k] = left_list[i]
            i += 1
        else:
            num_list[k] = right_list[j]
            j += 1
        k += 1

    while (i < len(left_list)):
        num_list[k] = left_list[i]
        i += 1
        k += 1

    while (j < len(right_list)):
        num_list[k] = right_list[j]
        j += 1
        k += 1

    return(num_list)


def quick_sort(data):
    if len(data) <= 1:
        return data

    pivot = data[0]

    left = [item for item in data[1:] if pivot > item]
    right = [item for item in data[1:] if pivot < item]

    return quick_sort(left) + [pivot] + quick_sort(right)


data_list = random.sample(range(100), 10)
print(quick_sort(data_list))